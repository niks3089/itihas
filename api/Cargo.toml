[package]
edition = "2021"
name = "api"
version = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
async-trait = {workspace = true}
anyhow = {workspace = true}
bs58 = {workspace = true}
cadence = {workspace = true}
cadence-macros = {workspace = true}
common = {workspace = true}
chrono = {workspace = true}
dao = {workspace = true}
env_logger = {workspace = true}
figment = {workspace = true}
hex = {workspace = true}
hyper = {workspace = true}
jsonrpsee = {workspace = true}
jsonrpsee-core = {workspace = true}
log = {workspace = true}
mockall = {workspace = true}
metrics = {workspace = true}
open-rpc-derive = {workspace = true}
open-rpc-schema = {workspace = true}
schemars = {workspace = true}
schemars_derive = {workspace = true}
sea-orm = {workspace = true}
serde = {workspace = true}
serde_json = {workspace = true}
solana-sdk = {workspace = true}
sqlx = {workspace = true}
thiserror = {workspace = true}
tokio = {workspace = true}
tokio-postgres = {workspace = true}
tower = {workspace = true}
tower-http = {workspace = true}
tracing = {workspace = true}
tracing-subscriber = { workspace = true }

[target.'cfg(not(target_env = "msvc"))'.dependencies]
tikv-jemallocator = "0.5"
