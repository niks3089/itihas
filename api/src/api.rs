use std::str::FromStr;

use crate::{
    db::{Dao, PageOptions, Pagination},
    error::ApiError,
    spec::TransactionList,
    types::Transaction,
};
use chrono::NaiveDate;
use common::db::setup_database_connection;
use solana_sdk::pubkey::Pubkey;

use crate::config::ApiConfig;

pub fn validate_pubkey(str_pubkey: String) -> Result<Pubkey, ApiError> {
    Pubkey::from_str(&str_pubkey).map_err(|_| ApiError::PubkeyValidationError(str_pubkey))
}

pub struct Api {
    pub config: ApiConfig,
    pub dao: Dao,
}

impl Api {
    pub async fn new(config: ApiConfig) -> Self {
        Api {
            config: config.clone(),
            dao: Dao::new(
                setup_database_connection(config.get_database_url(), config.max_connections)
                    .await
                    .into(),
            ),
        }
    }

    pub fn create_pagination(&self, page_opt: PageOptions) -> Result<Pagination, ApiError> {
        match (
            page_opt.before.as_ref(),
            page_opt.after.as_ref(),
            page_opt.page,
        ) {
            (_, _, None) => Ok(Pagination::Keyset {
                before: page_opt.before,
                after: page_opt.after,
            }),
            (None, None, Some(p)) => Ok(Pagination::Page { page: p }),
            _ => Err(ApiError::PaginationError),
        }
    }
    pub fn validate_pagination(
        &self,
        limit: &Option<u32>,
        page: &Option<u32>,
        before: &Option<String>,
        after: &Option<String>,
    ) -> Result<PageOptions, ApiError> {
        let mut page_opt = PageOptions::default();

        if let Some(limit) = limit {
            if *limit > 1000 {
                return Err(ApiError::PaginationExceededError);
            }
        }

        if let Some(page) = page {
            if *page == 0 {
                return Err(ApiError::PaginationEmptyError);
            }

            if before.is_some() || after.is_some() {
                return Err(ApiError::PaginationError);
            }

            let current_limit = limit.unwrap_or(1000);
            let offset = (*page - 1) * current_limit;
            if offset > 500_000 {
                return Err(ApiError::OffsetLimitExceededError);
            }
        }

        if let Some(before) = before {
            match NaiveDate::parse_from_str(before, "%d/%m/%Y") {
                Ok(date) => page_opt.before = Some(date),
                Err(_) => return Err(ApiError::InvalidDate("before".to_string())),
            }
        }

        if let Some(after) = after {
            match NaiveDate::parse_from_str(after, "%d/%m/%Y") {
                Ok(date) => page_opt.after = Some(date),
                Err(_) => return Err(ApiError::InvalidDate("after".to_string())),
            }
        }

        page_opt.limit = limit.map(|x| x as u64).unwrap_or(1000);
        page_opt.page = page.map(|x| x as u64);
        Ok(page_opt)
    }

    pub fn build_transaction_response(
        transactions: Vec<Transaction>,
        limit: u64,
        pagination: &Pagination,
    ) -> TransactionList {
        let total = transactions.len() as u32;
        let (page, before, after) = match pagination {
            Pagination::Keyset { before, after } => {
                let bef = before.map(|x| x.format("%d/%m/%Y").to_string());
                let aft = after.map(|x| x.format("%d/%m/%Y").to_string());
                (None, bef, aft)
            }
            Pagination::Page { page } => (Some(*page), None, None),
        };

        TransactionList {
            total,
            limit: limit as u32,
            page: page.map(|x| x as u32),
            before,
            after,
            items: transactions,
        }
    }
}
