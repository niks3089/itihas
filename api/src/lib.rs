pub mod api;
pub mod builder;
pub mod config;
pub mod db;
pub mod error;
pub mod spec;
pub mod types;
