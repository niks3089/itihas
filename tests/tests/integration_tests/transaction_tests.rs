use api::spec::{ApiContract, GetTransactionsByAddress};
use function_name::named;
use rstest::rstest;

use insta::assert_json_snapshot;
use serial_test::serial;

use crate::setup::cached_fetch_block;

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_get_transaction_by_address() {
    use crate::setup::{setup, trim_test_name, Network, TestSetupOptions};

    let name = trim_test_name(function_name!());
    let setup = setup(
        name.clone(),
        TestSetupOptions {
            network: Network::Mainnet,
        },
    )
    .await;

    let block = cached_fetch_block(&setup, 285941932).await;
    let _ = setup.dao.index_block(&block).await;
    let payload = GetTransactionsByAddress {
        source: Some("BhW85ig2dHu5tV6sCs7ps5UyPCLSjfXgEsfVAX82yXnb".to_string()),
        page: Some(1),
        destination: None,
        mint: None,
        before: None,
        after: None,
        limit: None,
        sort_by: None,
    };

    let parsed_transaction = setup
        .api
        .get_transactions_by_address(payload.clone())
        .await
        .unwrap();
    assert_json_snapshot!(
        format!("{}-{}-transaction-with-src", name.clone(), 1),
        parsed_transaction
    );

    let payload = GetTransactionsByAddress {
        source: Some("BhW85ig2dHu5tV6sCs7ps5UyPCLSjfXgEsfVAX82yXnb".to_string()),
        destination: Some("4HHZV2LRBQD5CJnYgMTzPeKcS2nnTT8szeh2svBWQ89m".to_string()),
        mint: None,
        page: Some(1),
        before: None,
        after: None,
        limit: None,
        sort_by: None,
    };

    let parsed_transaction = setup
        .api
        .get_transactions_by_address(payload.clone())
        .await
        .unwrap();
    assert_json_snapshot!(
        format!("{}-{}-transaction-with-dest", name.clone(), 1),
        parsed_transaction
    );

    let payload = GetTransactionsByAddress {
        source: Some("BhW85ig2dHu5tV6sCs7ps5UyPCLSjfXgEsfVAX82yXnb".to_string()),
        destination: None,
        mint: None,
        page: Some(2),
        before: None,
        after: None,
        limit: Some(5),
        sort_by: None,
    };

    let parsed_transaction = setup
        .api
        .get_transactions_by_address(payload.clone())
        .await
        .unwrap();
    assert_json_snapshot!(
        format!("{}-{}-transaction-with-page", name.clone(), 1),
        parsed_transaction
    );

    let payload = GetTransactionsByAddress {
        source: Some("HUe9Gfu8DMhY4Dj9A56N9muZg7euoFcXQskVAAfJpgEw".to_string()),
        destination: None,
        mint: None,
        page: Some(1),
        before: None,
        after: None,
        limit: None,
        sort_by: None,
    };

    let parsed_transaction = setup
        .api
        .get_transactions_by_address(payload.clone())
        .await
        .unwrap();
    assert_json_snapshot!(
        format!("{}-{}-transaction-unknown", name.clone(), 1),
        parsed_transaction
    );

    let payload = GetTransactionsByAddress {
        source: None,
        destination: None,
        mint: Some("AmeroCaeKg55p6J8d1y2R4t9taqgn3TH4BARgzQJyHvd".to_string()),
        page: Some(1),
        before: None,
        after: None,
        limit: None,
        sort_by: None,
    };

    let parsed_transaction = setup
        .api
        .get_transactions_by_address(payload.clone())
        .await
        .unwrap();
    assert_json_snapshot!(
        format!("{}-{}-transaction-mint", name.clone(), 1),
        parsed_transaction
    );
}
