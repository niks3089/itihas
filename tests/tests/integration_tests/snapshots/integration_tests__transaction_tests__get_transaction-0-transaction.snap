---
source: tests/tests/integration_tests/transaction_tests.rs
assertion_line: 43
expression: parsed_transaction
---
{
  "signature": "4gmMdnaUkDUySPtv7W3WdyPAnLtHrE4dfwkcZtGm9QCNxKNNx2RbwgATDyM95bS9V3GMbpKu3pab1XPcKvSgtDmN",
  "source_address": "5oTHwHSgkX2kMP88aRyy2Cvi8SqhRv7DJFbacRy9upCm",
  "token_type": "spl_token",
  "destination_address": "HUe9Gfu8DMhY4Dj9A56N9muZg7euoFcXQskVAAfJpgEw",
  "source_ata": "2xpg7vncUU7YCexwgKrAeWDSjk4of7T9UigPk6ChBo6s",
  "destination_ata": "2Wqzw7iES7vtKnZtkJaPXpcaq9Hpz3gDPKnqQipoDEct",
  "mint_address": "2smrziPxYKYJSECKjnAv4b6P4Qdz8B2SbkPt8F2Y4mH1",
  "slot": 317418346,
  "amount": 300000000,
  "block_time": "2024-08-08T17:53:33Z"
}
